File: src/resolver.ts
// Name resolution: map a name-use Identifier to the Symbol it refers to.
//
// Lexical scope chain (JLS 6.5, simplified): block/method locals + type
// parameters -> type members (incl. inherited from super types) -> enclosing
// types -> single-type imports -> same package -> on-demand imports + java.lang
// -> fully-qualified names via the program's global index. Member access of an
// expression (a.b) needs the checker (P5) and is deferred.

import { forEachChild } from "./parser.ts";
import { type Uri } from "./workspace.ts";
import type { Fqn, GlobalIndex, PackageName, Program } from "./program.ts";
import {
  type ClassDeclaration,
  type EntityName,
  type EnumDeclaration,
  type Identifier,
  type InterfaceDeclaration,
  type Node,
  type ObjectCreationExpression,
  type RecordDeclaration,
  type SourceFile,
  type Symbol,
  SymbolFlags,
  SyntaxKind,
  type TypeNode,
  type TypeReference,
} from "./types.ts";
import { entityNameToString } from "./utilities.ts";

export const enum Meaning {
  Any,
  Type,
  Value,
}

function isTypeDeclaration(node: Node): boolean {
  switch (node.kind) {
    case SyntaxKind.ClassDeclaration:
    case SyntaxKind.InterfaceDeclaration:
    case SyntaxKind.EnumDeclaration:
    case SyntaxKind.AnnotationTypeDeclaration:
    case SyntaxKind.RecordDeclaration:
      return true;
    default:
      return false;
  }
}

function matchesMeaning(symbol: Symbol, meaning: Meaning): boolean {
  switch (meaning) {
    case Meaning.Type:
      return (symbol.flags & SymbolFlags.Type) !== 0;
    case Meaning.Value:
      return (
        (symbol.flags &
          (SymbolFlags.Field |
            SymbolFlags.Parameter |
            SymbolFlags.LocalVariable |
            SymbolFlags.EnumConstant |
            SymbolFlags.Method)) !==
        0
      );
    default:
      return true;
  }
}

export function getSourceFileOfNode(node: Node): SourceFile {
  let current = node;
  while (current.kind !== SyntaxKind.SourceFile) {
    current = current.parent;
  }
  return current as SourceFile;
}

function packageNameOf(sourceFile: SourceFile): PackageName {
  return (
    sourceFile.packageDeclaration ? entityNameToString(sourceFile.packageDeclaration.name) : ""
  ) as PackageName;
}

function lastSegment(qualified: string): string {
  const dot = qualified.lastIndexOf(".");
  return dot < 0 ? qualified : qualified.slice(dot + 1);
}

// --- inheritance-aware member lookup -------------------------------------------------

// Guards against cycles in the super-type graph while resolving inherited members.
const resolvingSupertypes = new Set<Symbol>();

function superTypeNodes(declaration: Node): readonly TypeNode[] {
  switch (declaration.kind) {
    case SyntaxKind.ClassDeclaration: {
      const c = declaration as ClassDeclaration;
      return [...(c.extendsType ? [c.extendsType] : []), ...(c.implementsTypes ?? [])];
    }
    case SyntaxKind.InterfaceDeclaration:
      return (declaration as InterfaceDeclaration).extendsTypes ?? [];
    case SyntaxKind.EnumDeclaration:
      return (declaration as EnumDeclaration).implementsTypes ?? [];
    case SyntaxKind.RecordDeclaration:
      return (declaration as RecordDeclaration).implementsTypes ?? [];
    default:
      return [];
  }
}

/** Direct super-type symbols (extends/implements) of a type symbol. */
export function getDirectSuperTypeSymbols(typeSymbol: Symbol, program: Program): Symbol[] {
  return superTypeSymbols(typeSymbol, program);
}

function superTypeSymbols(typeSymbol: Symbol, program: Program): Symbol[] {
  if (resolvingSupertypes.has(typeSymbol)) return [];
  const declaration = typeSymbol.valueDeclaration ?? typeSymbol.declarations?.[0];
  if (!declaration) return [];
  resolvingSupertypes.add(typeSymbol);
  try {
    const result: Symbol[] = [];
    // An enum implicitly extends java.lang.Enum (JLS 8.9): name(), ordinal(),
    // compareTo(), etc.
    if (declaration.kind === SyntaxKind.EnumDeclaration) {
      const enumSymbol = program.getGlobalIndex().getType("java.lang.Enum" as Fqn);
      if (enumSymbol) result.push(enumSymbol);
    }
    for (const typeNode of superTypeNodes(declaration)) {
      if (typeNode.kind === SyntaxKind.TypeReference) {
        const symbol = resolveTypeEntityName(
          (typeNode as TypeReference).typeName,
          declaration,
          program,
        );
        if (symbol) result.push(symbol);
      }
    }
    return result;
  } finally {
    resolvingSupertypes.delete(typeSymbol);
  }
}

/** Look up a member by name in a type and its super types. */
export function lookupMember(
  typeSymbol: Symbol,
  name: string,
  meaning: Meaning,
  program: Program,
  seen = new Set<Symbol>(),
): Symbol | undefined {
  if (seen.has(typeSymbol)) return undefined;
  seen.add(typeSymbol);
  const own = typeSymbol.members?.get(name);
  if (own && matchesMeaning(own, meaning)) return own;
  for (const superSymbol of superTypeSymbols(typeSymbol, program)) {
    const inherited = lookupMember(superSymbol, name, meaning, program, seen);
    if (inherited) return inherited;
  }
  return undefined;
}

// --- scope chain ----------------------------------------------------------------------

function lookupInScopes(
  start: Node,
  name: string,
  meaning: Meaning,
  program: Program,
): Symbol | undefined {
  let node: Node | undefined = start;
  let prev: Node | undefined;
  while (node) {
    if (isTypeDeclaration(node) && node.symbol) {
      const member = lookupMember(node.symbol, name, meaning, program);
      if (member) return member;
    } else if (
      node.kind === SyntaxKind.ObjectCreationExpression &&
      (node as ObjectCreationExpression).classBody &&
      prev !== undefined &&
      ((node as ObjectCreationExpression).classBody as readonly Node[]).includes(prev)
    ) {
      // Inside an anonymous class body: members are inherited from the type it
      // extends/implements, so look them up on the supertype.
      const oce = node as ObjectCreationExpression;
      const target =
        oce.type.kind === SyntaxKind.TypeReference
          ? resolveTypeEntityName((oce.type as TypeReference).typeName, oce, program)
          : undefined;
      if (target) {
        const member = lookupMember(target, name, meaning, program);
        if (member) return member;
      }
    } else {
      const local = node.locals?.get(name);
      if (local && matchesMeaning(local, meaning)) return local;
    }
    prev = node;
    node = node.parent;
  }
  return undefined;
}

// --- cross-file type resolution -------------------------------------------------------

function resolveTypeNameCrossFile(
  name: string,
  sourceFile: SourceFile,
  index: GlobalIndex,
): Symbol | undefined {
  // single-type imports
  for (const imp of sourceFile.imports) {
    if (!imp.isStatic && !imp.isOnDemand) {
      const fqn = entityNameToString(imp.name) as Fqn;
      if (lastSegment(fqn) === name) {
        const type = index.getType(fqn);
        if (type) return type;
      }
    }
  }
  // same package
  const samePackage = index.getPackageTypes(packageNameOf(sourceFile))?.get(name);
  if (samePackage) return samePackage;
  // on-demand imports
  for (const imp of sourceFile.imports) {
    if (!imp.isStatic && imp.isOnDemand) {
      const type = index.getPackageTypes(entityNameToString(imp.name) as PackageName)?.get(name);
      if (type) return type;
    }
  }
  // implicit java.lang.*
  return index.getPackageTypes("java.lang" as PackageName)?.get(name);
}

// A value/method imported via `import static T.member` or `import static T.*`
// (JLS 7.5.3/7.5.4): resolve the named type and look the member up on it.
function resolveStaticImport(
  name: string,
  sourceFile: SourceFile,
  program: Program,
): Symbol | undefined {
  const index = program.getGlobalIndex();
  for (const imp of sourceFile.imports) {
    if (!imp.isStatic) continue;
    const fqn = entityNameToString(imp.name) as Fqn;
    if (imp.isOnDemand) {
      const type = index.getType(fqn);
      const member = type && lookupMember(type, name, Meaning.Any, program);
      if (member) return member;
    } else if (lastSegment(fqn) === name) {
      const type = index.getType(fqn.slice(0, fqn.lastIndexOf(".")) as Fqn);
      const member = type && lookupMember(type, name, Meaning.Any, program);
      if (member) return member;
    }
  }
  return undefined;
}

function resolveTypeName(name: string, fromNode: Node, program: Program): Symbol | undefined {
  const lexical = lookupInScopes(fromNode, name, Meaning.Type, program);
  if (lexical) return lexical;
  return resolveTypeNameCrossFile(name, getSourceFileOfNode(fromNode), program.getGlobalIndex());
}

// Per-node resolution memo (the TypeScript compiler's nodeLinks pattern): a
// reparse creates fresh nodes, so stale entries die with their keys. `null`
// records a resolved-to-nothing answer (distinct from "not yet computed").
const typeNameLinks = new WeakMap<Node, Symbol | null>();

/** Resolve an entity name used as a type (Identifier, qualified, or nested). */
export function resolveTypeEntityName(
  name: EntityName,
  fromNode: Node,
  program: Program,
): Symbol | undefined {
  const cached = typeNameLinks.get(name);
  if (cached !== undefined) return cached ?? undefined;
  const result = resolveTypeEntityNameWorker(name, fromNode, program);
  typeNameLinks.set(name, result ?? null);
  return result;
}

function resolveTypeEntityNameWorker(
  name: EntityName,
  fromNode: Node,
  program: Program,
): Symbol | undefined {
  if (name.kind === SyntaxKind.Identifier) {
    return resolveTypeName((name as Identifier).text, fromNode, program);
  }
  const fqn = entityNameToString(name) as Fqn;
  const byFqn = program.getGlobalIndex().getType(fqn);
  if (byFqn) return byFqn;
  // nested type: resolve the left as a type, then look up the right member type
  const qualified = name as { left: EntityName; right: Identifier };
  const leftType = resolveTypeEntityName(qualified.left, fromNode, program);
  if (leftType) return lookupMember(leftType, qualified.right.text, Meaning.Type, program);
  return undefined;
}

// --- identifier classification --------------------------------------------------------

function declarationOf(identifier: Identifier): Node | undefined {
  const parent = identifier.parent;
  if (parent && parent.symbol && (parent as { name?: Node }).name === identifier) return parent;
  return undefined;
}

function isQualifiedTypeNameTail(identifier: Identifier): boolean {
  const parent = identifier.parent;
  return (
    !!parent &&
    parent.kind === SyntaxKind.QualifiedName &&
    (parent as { right?: Node }).right === identifier &&
    meaningOf(identifier) === Meaning.Type
  );
}

function isExpressionMemberAccess(identifier: Identifier): boolean {
  const parent = identifier.parent;
  if (!parent) return false;
  return (
    (parent.kind === SyntaxKind.PropertyAccessExpression ||
      parent.kind === SyntaxKind.MethodReferenceExpression) &&
    (parent as { name?: Node }).name === identifier
  );
}

function meaningOf(identifier: Identifier): Meaning {
  let node: Node | undefined = identifier;
  while (node) {
    if (node.kind === SyntaxKind.TypeReference) return Meaning.Type;
    if (node.kind !== SyntaxKind.QualifiedName && node.kind !== SyntaxKind.Identifier) break;
    node = node.parent;
  }
  return Meaning.Any;
}

/** Resolve a name-use identifier to its declaration symbol, or undefined. */
const identifierLinks = new WeakMap<Node, Symbol | null>();

export function resolveIdentifier(identifier: Identifier, program: Program): Symbol | undefined {
  const cached = identifierLinks.get(identifier);
  if (cached !== undefined) return cached ?? undefined;
  const result = resolveIdentifierWorker(identifier, program);
  identifierLinks.set(identifier, result ?? null);
  return result;
}

function resolveIdentifierWorker(identifier: Identifier, program: Program): Symbol | undefined {
  const declaration = declarationOf(identifier);
  if (declaration) return declaration.symbol;

  if (isQualifiedTypeNameTail(identifier)) {
    return resolveTypeEntityName(identifier.parent as EntityName, identifier, program);
  }
  if (isExpressionMemberAccess(identifier)) return undefined; // needs the checker (P5)

  const meaning = meaningOf(identifier);
  const lexical = lookupInScopes(identifier, identifier.text, meaning, program);
  if (lexical) return lexical;
  if (meaning !== Meaning.Value) {
    const type = resolveTypeNameCrossFile(
      identifier.text,
      getSourceFileOfNode(identifier),
      program.getGlobalIndex(),
    );
    if (type) return type;
  }
  // A statically-imported field or method used by its simple name.
  return resolveStaticImport(identifier.text, getSourceFileOfNode(identifier), program);
}

export function getDeclarationNameNode(symbol: Symbol): Node | undefined {
  const declaration = symbol.valueDeclaration ?? symbol.declarations?.[0];
  if (!declaration) return undefined;
  return (declaration as { name?: Node }).name ?? declaration;
}

// --- find references ------------------------------------------------------------------

function forEachDescendant(node: Node, cb: (n: Node) => void): void {
  cb(node);
  forEachChild(node, child => {
    forEachDescendant(child, cb);
    return undefined;
  });
}

// Locals, parameters and type parameters cannot be referenced outside the file
// that declares them, so narrow the search to that file. Everything else (types,
// fields, methods) may be referenced cross-file, so the whole workspace is scanned.
const FILE_LOCAL_FLAGS =
  SymbolFlags.LocalVariable | SymbolFlags.Parameter | SymbolFlags.TypeParameter;

function candidateUris(symbol: Symbol, program: Program): Uri[] {
  if (symbol.flags & FILE_LOCAL_FLAGS) {
    const declaration = symbol.valueDeclaration ?? symbol.declarations?.[0];
    // a SourceFile's fileName is the uri it was registered under
    if (declaration) return [getSourceFileOfNode(declaration).fileName as Uri];
  }
  return program.getAllUris();
}

/**
 * All identifier nodes (uses and declaration names) that refer to a symbol.
 *
 * `resolve` maps an identifier to its symbol. The default resolves lexical names
 * only; pass the checker's resolveName to also match member accesses (a.field),
 * which is required for a correct rename of fields and methods.
 */
export function findReferences(
  symbol: Symbol,
  program: Program,
  resolve: (id: Identifier) => Symbol | undefined = id => resolveIdentifier(id, program),
): Node[] {
  const result: Node[] = [];
  for (const uri of candidateUris(symbol, program)) {
    const sourceFile = program.getSourceFile(uri);
    if (!sourceFile) continue;
    forEachDescendant(sourceFile, node => {
      if (node.kind !== SyntaxKind.Identifier) return;
      if (resolve(node as Identifier) === symbol) result.push(node);
    });
  }
  return result;
}

File: togo/internal/compiler/resolver.go
package compiler

// Name resolution: map a name-use Identifier to the Symbol it refers to.
//
// Lexical scope chain (JLS 6.5, simplified): block/method locals + type
// parameters -> type members (incl. inherited from super types) -> enclosing
// types -> single-type imports -> same package -> on-demand imports + java.lang
// -> fully-qualified names via the program's global index. Member access of an
// expression (a.b) needs the checker (P5) and is deferred.
// Port of src/compiler/resolver.ts.

import "strings"

type Meaning int

const (
	MeaningAny Meaning = iota
	MeaningType
	MeaningValue
)

func resolverIsTypeDeclaration(node *Node) bool {
	switch node.Kind {
	case ClassDeclaration, InterfaceDeclaration, EnumDeclaration, AnnotationTypeDeclaration, RecordDeclaration:
		return true
	default:
		return false
	}
}

func matchesMeaning(symbol *Symbol, meaning Meaning) bool {
	switch meaning {
	case MeaningType:
		return symbol.Flags&SymbolFlagsType != 0
	case MeaningValue:
		return symbol.Flags&(SymbolFlagsField|SymbolFlagsParameter|SymbolFlagsLocalVariable|SymbolFlagsEnumConstant|SymbolFlagsMethod) != 0
	default:
		return true
	}
}

// GetSourceFileOfNode walks up to the enclosing SourceFile node.
func GetSourceFileOfNode(node *Node) *Node {
	current := node
	for current.Kind != SourceFile {
		current = current.Parent
	}
	return current
}

func packageNameOf(sourceFile *Node) PackageName {
	data := sourceFile.AsSourceFile()
	if data.PackageDeclaration != nil {
		return PackageName(entityNameToString(data.PackageDeclaration.AsPackageDeclaration().Name))
	}
	return ""
}

func lastSegment(qualified string) string {
	dot := strings.LastIndex(qualified, ".")
	if dot < 0 {
		return qualified
	}
	return qualified[dot+1:]
}

// --- inheritance-aware member lookup -----------------------------------------

// resolvingSupertypes guards against cycles in the super-type graph.
var resolvingSupertypes = map[*Symbol]bool{}

func superTypeNodes(declaration *Node) []*Node {
	switch declaration.Kind {
	case ClassDeclaration:
		c := declaration.AsClassDeclaration()
		var out []*Node
		if c.ExtendsType != nil {
			out = append(out, c.ExtendsType)
		}
		if c.ImplementsTypes != nil {
			out = append(out, c.ImplementsTypes.Nodes...)
		}
		return out
	case InterfaceDeclaration:
		if e := declaration.AsInterfaceDeclaration().ExtendsTypes; e != nil {
			return e.Nodes
		}
	case EnumDeclaration:
		if i := declaration.AsEnumDeclaration().ImplementsTypes; i != nil {
			return i.Nodes
		}
	case RecordDeclaration:
		if i := declaration.AsRecordDeclaration().ImplementsTypes; i != nil {
			return i.Nodes
		}
	}
	return nil
}

// GetDirectSuperTypeSymbols returns the direct super-type symbols (extends/implements).
func GetDirectSuperTypeSymbols(typeSymbol *Symbol, program *Program) []*Symbol {
	return superTypeSymbols(typeSymbol, program)
}

func superTypeSymbols(typeSymbol *Symbol, program *Program) []*Symbol {
	if resolvingSupertypes[typeSymbol] {
		return nil
	}
	declaration := declarationOfSymbol(typeSymbol)
	if declaration == nil {
		return nil
	}
	resolvingSupertypes[typeSymbol] = true
	defer delete(resolvingSupertypes, typeSymbol)

	var result []*Symbol
	// An enum implicitly extends java.lang.Enum (JLS 8.9).
	if declaration.Kind == EnumDeclaration {
		if enumSymbol := program.GetGlobalIndex().GetType("java.lang.Enum"); enumSymbol != nil {
			result = append(result, enumSymbol)
		}
	}
	for _, typeNode := range superTypeNodes(declaration) {
		if typeNode.Kind == TypeReference {
			if symbol := ResolveTypeEntityName(typeNode.AsTypeReference().TypeName, declaration, program); symbol != nil {
				result = append(result, symbol)
			}
		}
	}
	// Every class without an extends clause (and every record) implicitly extends
	// java.lang.Object (JLS 8.1.4); an interface's members include Object's public
	// ones (JLS 9.2).
	hasClassSuper := declaration.Kind == ClassDeclaration && declaration.AsClassDeclaration().ExtendsType != nil
	if !hasClassSuper {
		if objectSymbol := program.GetGlobalIndex().GetType("java.lang.Object"); objectSymbol != nil && objectSymbol != typeSymbol {
			result = append(result, objectSymbol)
		}
	}
	return result
}

func declarationOfSymbol(symbol *Symbol) *Node {
	if symbol.ValueDeclaration != nil {
		return symbol.ValueDeclaration
	}
	if len(symbol.Declarations) > 0 {
		return symbol.Declarations[0]
	}
	return nil
}

// LookupMember looks up a member by name in a type and its super types.
func LookupMember(typeSymbol *Symbol, name string, meaning Meaning, program *Program) *Symbol {
	return lookupMember(typeSymbol, name, meaning, program, map[*Symbol]bool{})
}

func lookupMember(typeSymbol *Symbol, name string, meaning Meaning, program *Program, seen map[*Symbol]bool) *Symbol {
	if seen[typeSymbol] {
		return nil
	}
	seen[typeSymbol] = true
	if own := typeSymbol.Members[name]; own != nil && matchesMeaning(own, meaning) {
		return own
	}
	for _, superSymbol := range superTypeSymbols(typeSymbol, program) {
		if inherited := lookupMember(superSymbol, name, meaning, program, seen); inherited != nil {
			return inherited
		}
	}
	return nil
}

// --- scope chain -------------------------------------------------------------

func lookupInScopes(start *Node, name string, meaning Meaning, program *Program) *Symbol {
	var node, prev *Node
	node = start
	for node != nil {
		switch {
		case resolverIsTypeDeclaration(node) && node.Symbol != nil:
			if member := LookupMember(node.Symbol, name, meaning, program); member != nil {
				return member
			}
		case node.Kind == ObjectCreationExpression && classBodyIncludes(node, prev):
			// Inside an anonymous class body: members are inherited from the type it
			// extends/implements, so look them up on the supertype.
			oce := node.AsObjectCreationExpression()
			var target *Symbol
			if oce.Type.Kind == TypeReference {
				target = ResolveTypeEntityName(oce.Type.AsTypeReference().TypeName, node, program)
			}
			if target != nil {
				if member := LookupMember(target, name, meaning, program); member != nil {
					return member
				}
			}
		default:
			if local := node.Locals[name]; local != nil && matchesMeaning(local, meaning) {
				return local
			}
		}
		prev = node
		node = node.Parent
	}
	return nil
}

func classBodyIncludes(node, prev *Node) bool {
	if prev == nil {
		return false
	}
	body := node.AsObjectCreationExpression().ClassBody
	if body == nil {
		return false
	}
	for _, m := range body.Nodes {
		if m == prev {
			return true
		}
	}
	return false
}

// --- cross-file type resolution ----------------------------------------------

func resolveTypeNameCrossFile(name string, sourceFile *Node, index *GlobalIndex) *Symbol {
	data := sourceFile.AsSourceFile()
	// single-type imports
	for _, imp := range data.Imports.Nodes {
		d := imp.AsImportDeclaration()
		if !d.IsStatic && !d.IsOnDemand {
			fqn := entityNameToString(d.Name)
			if lastSegment(fqn) == name {
				if t := index.GetType(Fqn(fqn)); t != nil {
					return t
				}
			}
		}
	}
	// same package
	if pkg := index.GetPackageTypes(packageNameOf(sourceFile)); pkg != nil {
		if samePackage := pkg[name]; samePackage != nil {
			return samePackage
		}
	}
	// on-demand imports. The package-types map only covers project/classpath
	// types; GetType also reaches the lazy JDK provider for a star-imported JDK
	// package (e.g. `import java.util.*` -> List).
	for _, imp := range data.Imports.Nodes {
		d := imp.AsImportDeclaration()
		if !d.IsStatic && d.IsOnDemand {
			pkgName := entityNameToString(d.Name)
			if pkg := index.GetPackageTypes(PackageName(pkgName)); pkg != nil {
				if t := pkg[name]; t != nil {
					return t
				}
			}
			if t := index.GetType(Fqn(pkgName + "." + name)); t != nil {
				return t
			}
		}
	}
	// implicit java.lang.* (GetType reaches the lazy JDK provider too)
	if pkg := index.GetPackageTypes("java.lang"); pkg != nil {
		if t := pkg[name]; t != nil {
			return t
		}
	}
	return index.GetType(Fqn("java.lang." + name))
}

// resolveStaticImport resolves a value/method imported via `import static`.
func resolveStaticImport(name string, sourceFile *Node, program *Program) *Symbol {
	index := program.GetGlobalIndex()
	for _, imp := range sourceFile.AsSourceFile().Imports.Nodes {
		d := imp.AsImportDeclaration()
		if !d.IsStatic {
			continue
		}
		fqn := entityNameToString(d.Name)
		if d.IsOnDemand {
			t := index.GetType(Fqn(fqn))
			if t != nil {
				if member := LookupMember(t, name, MeaningAny, program); member != nil {
					return member
				}
			}
		} else if lastSegment(fqn) == name {
			dot := strings.LastIndex(fqn, ".")
			if dot >= 0 {
				t := index.GetType(Fqn(fqn[:dot]))
				if t != nil {
					if member := LookupMember(t, name, MeaningAny, program); member != nil {
						return member
					}
				}
			}
		}
	}
	return nil
}

func resolveTypeName(name string, fromNode *Node, program *Program) *Symbol {
	if lexical := lookupInScopes(fromNode, name, MeaningType, program); lexical != nil {
		return lexical
	}
	return resolveTypeNameCrossFile(name, GetSourceFileOfNode(fromNode), program.GetGlobalIndex())
}

// typeNameLinks memoizes type-entity-name resolution per node (the nodeLinks
// pattern). A reparse creates fresh nodes, so stale entries die with their keys.
// A recorded resolvedNothing distinguishes "resolved to nothing" from "not computed".
var typeNameLinks = map[*Node]resolvedLink{}

type resolvedLink struct {
	symbol   *Symbol
	computed bool
}

// ResolveTypeEntityName resolves an entity name used as a type.
func ResolveTypeEntityName(name *Node, fromNode *Node, program *Program) *Symbol {
	if cached, ok := typeNameLinks[name]; ok && cached.computed {
		return cached.symbol
	}
	result := resolveTypeEntityNameWorker(name, fromNode, program)
	typeNameLinks[name] = resolvedLink{symbol: result, computed: true}
	return result
}

func resolveTypeEntityNameWorker(name *Node, fromNode *Node, program *Program) *Symbol {
	if name.Kind == Identifier {
		return resolveTypeName(name.AsIdentifier().Text, fromNode, program)
	}
	fqn := entityNameToString(name)
	if byFqn := program.GetGlobalIndex().GetType(Fqn(fqn)); byFqn != nil {
		return byFqn
	}
	// nested type: resolve the left as a type, then look up the right member type
	q := name.AsQualifiedName()
	if leftType := ResolveTypeEntityName(q.Left, fromNode, program); leftType != nil {
		return LookupMember(leftType, q.Right.AsIdentifier().Text, MeaningType, program)
	}
	return nil
}

// --- identifier classification -----------------------------------------------

func declarationOf(identifier *Node) *Node {
	parent := identifier.Parent
	if parent != nil && parent.Symbol != nil && nodeName(parent) == identifier {
		return parent
	}
	return nil
}

func isQualifiedTypeNameTail(identifier *Node) bool {
	parent := identifier.Parent
	return parent != nil && parent.Kind == QualifiedName &&
		parent.AsQualifiedName().Right == identifier &&
		meaningOf(identifier) == MeaningType
}

func isExpressionMemberAccess(identifier *Node) bool {
	parent := identifier.Parent
	if parent == nil {
		return false
	}
	if parent.Kind == PropertyAccessExpression {
		return parent.AsPropertyAccessExpression().Name == identifier
	}
	if parent.Kind == MethodReferenceExpression {
		return parent.AsMethodReferenceExpression().Name == identifier
	}
	return false
}

func meaningOf(identifier *Node) Meaning {
	node := identifier
	for node != nil {
		if node.Kind == TypeReference {
			return MeaningType
		}
		if node.Kind != QualifiedName && node.Kind != Identifier {
			break
		}
		node = node.Parent
	}
	return MeaningAny
}

// identifierLinks memoizes identifier resolution per node.
var identifierLinks = map[*Node]resolvedLink{}

// ResolveIdentifier resolves a name-use identifier to its declaration symbol, or nil.
func ResolveIdentifier(identifier *Node, program *Program) *Symbol {
	if cached, ok := identifierLinks[identifier]; ok && cached.computed {
		return cached.symbol
	}
	result := resolveIdentifierWorker(identifier, program)
	identifierLinks[identifier] = resolvedLink{symbol: result, computed: true}
	return result
}

func resolveIdentifierWorker(identifier *Node, program *Program) *Symbol {
	if declaration := declarationOf(identifier); declaration != nil {
		return declaration.Symbol
	}
	if isQualifiedTypeNameTail(identifier) {
		return ResolveTypeEntityName(identifier.Parent, identifier, program)
	}
	if isExpressionMemberAccess(identifier) {
		return nil // needs the checker (P5)
	}
	meaning := meaningOf(identifier)
	if lexical := lookupInScopes(identifier, identifier.AsIdentifier().Text, meaning, program); lexical != nil {
		return lexical
	}
	if meaning != MeaningValue {
		if t := resolveTypeNameCrossFile(identifier.AsIdentifier().Text, GetSourceFileOfNode(identifier), program.GetGlobalIndex()); t != nil {
			return t
		}
	}
	// A statically-imported field or method used by its simple name.
	return resolveStaticImport(identifier.AsIdentifier().Text, GetSourceFileOfNode(identifier), program)
}

// GetDeclarationNameNode returns the name node of a symbol's declaration.
func GetDeclarationNameNode(symbol *Symbol) *Node {
	declaration := declarationOfSymbol(symbol)
	if declaration == nil {
		return nil
	}
	if name := nodeName(declaration); name != nil {
		return name
	}
	return declaration
}

// --- find references ---------------------------------------------------------

func forEachDescendant(node *Node, cb func(*Node)) {
	cb(node)
	node.ForEachChild(func(child *Node) bool {
		forEachDescendant(child, cb)
		return false
	})
}

// fileLocalFlags marks symbols that cannot be referenced outside their file.
const fileLocalFlags = SymbolFlagsLocalVariable | SymbolFlagsParameter | SymbolFlagsTypeParameter

func candidateUris(symbol *Symbol, program *Program) []URI {
	if symbol.Flags&fileLocalFlags != 0 {
		if declaration := declarationOfSymbol(symbol); declaration != nil {
			return []URI{URI(GetSourceFileOfNode(declaration).AsSourceFile().FileName)}
		}
	}
	return program.GetAllUris()
}

// FindReferences returns all identifier nodes (uses and declaration names) that
// refer to a symbol. resolve maps an identifier to its symbol; pass nil to use
// the default lexical resolver.
func FindReferences(symbol *Symbol, program *Program, resolve func(*Node) *Symbol) []*Node {
	if resolve == nil {
		resolve = func(id *Node) *Symbol { return ResolveIdentifier(id, program) }
	}
	var result []*Node
	for _, uri := range candidateUris(symbol, program) {
		sourceFile := program.GetSourceFile(uri)
		if sourceFile == nil {
			continue
		}
		forEachDescendant(sourceFile, func(node *Node) {
			if node.Kind != Identifier {
				return
			}
			if resolve(node) == symbol {
				result = append(result, node)
			}
		})
	}
	return result
}
